Fix: Replace D_ASSERT(index.IsBound()) with proper error handling to prevent crashes during WAL replay

arghaffari opened the pull request:
This PR addresses issue duckdb/duckdb-rs#649 where DuckDB crashes with an assertion failure during corrupted WAL replay.

   ## Problem
   When opening a DuckDB database with a corrupted or incomplete WAL file, DuckDB crashes due to `D_ASSERT(index.IsBound())` assertions, causing `abort()`. This is unrecoverable at the application level.

   ## Solution
   Replace `D_ASSERT(index.IsBound())` with proper error handling:
   - Skip unbound indexes where safe (index verification, conflict detection, scanning)
   - Throw `SerializationException` for WAL-related paths (enables recovery via truncation)

   ## Files Changed
   - `src/storage/data_table.cpp`
   - `src/storage/write_ahead_log.cpp`
   - `src/storage/table_index_list.cpp`
   - `src/storage/local_storage.cpp`
   - `src/function/table/table_scan.cpp`
   - `src/execution/operator/persistent/physical_insert.cpp`
   - `test/persistence/test_corrupted_wal_index.cpp` (new test)

   Fixes duckdb/duckdb-rs#649

artjomPlaunov commented:
@arghaffari is there a database file to go along with the WAL we could take a look at, i.e. the database along with WAL so we can see what is going on when it starts up

arghaffari commented:
@artjomPlaunov 
[Archive 2.zip](https://github.com/user-attachments/files/24317883/Archive.2.zip)

iaintbyte committed:
Fix: Bind all table indexes after WAL replay to prevent crashes

## Fundamental Fix for Issue duckdb/duckdb-rs#649

This commit addresses the root cause of crashes when opening databases with
corrupted or truncated WAL files.

### Root Cause Analysis:
1. During WAL replay, indexes are created as UnboundIndex objects
2. These are stored in replay_index_infos and only added to tables on WAL_FLUSH
3. If WAL is truncated before FLUSH, indexes may remain unbound
4. Later operations crash when they encounter unbound indexes (D_ASSERT fails)

### Solution:
Add a post-replay step that binds all table indexes after WAL replay completes.
This ensures database consistency regardless of whether WAL replay was
complete or partial.

The new BindAllTableIndexesAfterReplay() function:
- Iterates through all schemas and tables in the catalog
- Calls BindIndexes() on each table's storage
- Gracefully handles binding failures to prevent cascade crashes
- Ensures all indexes are in a consistent state before the database is used

### Why This Works:
- BindIndexes() is the normal path for binding unbound indexes
- It's already called before every operation that needs bound indexes
- By calling it at the end of WAL replay, we ensure the database is consistent
- Even if binding fails for some indexes, the database remains usable
  (operations on those specific indexes will throw proper errors)

### Benefits:
- Fixes the root cause rather than adding safety checks everywhere
- Minimal code change (single function addition + one call site)
- Follows existing DuckDB patterns for index binding
- Graceful degradation - database opens even with partial corruption

iaintbyte committed:
Revert fundamental fix, use targeted safety checks instead

- Remove BindAllTableIndexesAfterReplay function (defeats lazy binding)
- Keep D_ASSERT(index.IsBound()) for debug builds
- Add if (!index.IsBound()) throw InternalException for release builds
- Preserves DuckDB's lazy binding performance optimization
- Targeted fixes at each D_ASSERT location rather than global startup binding

arghaffari commented:
@artjomPlaunov Yes, exactly. Here's the scenario:

### Our Setup
- **Tauri desktop app** (macOS/Windows/Linux)
- Using **duckdb-rs** for local data storage
- Tables have `PRIMARY KEY`, `UNIQUE`, and `FOREIGN KEY` constraints

### The Crash Scenario

1. **User is actively using the app** - writes happening (INSERTs/UPDATEs to indexed tables)

2. **User force-quits** via:
   - `Cmd+Q` on macOS
   - `Alt+F4` on Windows  
   - Or system kill/crash

3. **DuckDB doesn't get clean shutdown** - WAL file left mid-write

4. **User reopens app** - DuckDB tries to replay WAL → crash on `D_ASSERT(index.IsBound())`

### Why Force-Quit Causes Truncated WAL

When the process is killed mid-write:
- OS may not flush pending writes to disk
- WAL file ends up with incomplete entries
- Index creation/update entries are partially written
- On replay, the index metadata exists but the actual index data isn't bound

### Our Current Workaround

```rust
// Before opening DuckDB, delete any existing WAL
let wal_path = format!("{}.wal", db_path);
if std::path::Path::new(&wal_path).exists() {
    std::fs::remove_file(&wal_path)?;
}
```

This prevents crashes but loses uncommitted data. We'd prefer DuckDB to either:
1. Gracefully skip the corrupted portion and recover what it can
2. Throw a catchable exception so we can handle it in the app

Does this help clarify the scenario?

artjomPlaunov commented:
i.e. checkout the commit for the tag 1.4.3, make debug, then run the test using the debug unittest.

arghaffari commented:
@artjomPlaunov Final update from our investigation:

### Testing Summary

I tried multiple WAL corruption scenarios:
1. **Force-quit during writes** → Recovered gracefully
2. **Truncated WAL to 60%** → Recovered gracefully  
3. **Zeroed bytes 5000-15000** → Recovered gracefully

**Result**: Cannot reproduce the D_ASSERT crash with duckdb-rs 1.4.

### Root Cause Analysis

The reason we can't reproduce it:

1. **duckdb-rs uses `bundled` feature** which includes a **release build** of DuckDB
2. `D_ASSERT` macros are **compiled out** in release builds
3. In release mode, DuckDB throws `CatalogException` or silently recovers

### When Would D_ASSERT Fire?

Only when:
1. Building DuckDB from source with `-DCMAKE_BUILD_TYPE=Debug`
2. Running the test against that debug build

### Recommendation

Since:
- Release builds handle corrupted WAL gracefully (no crash)
- Debug builds are only used by developers
- The fix in this PR would only affect debug build behavior

I think this PR could be:
1. **Closed** if the current behavior (graceful handling in release, assertion in debug) is acceptable
2. **Kept** if you want debug builds to also handle this gracefully (better developer experience)

What do you think? Should I close this PR, or is there value in having debug builds also throw exceptions instead of asserting?
